bcmp/dfu_host.c

```diff
@@ -159,9 +161,18 @@ static BmErr bm_dfu_host_send_chunk(BmDfuEventChunkRequest *req) {
                             host_ctx.img_info.image_size -
                             host_ctx.bytes_remaining;
     do {
-      err =
-          bm_dfu_host_get_chunk(flash_offset, payload_header->chunk.payload_buf,
-                                payload_len, flash_read_timeout_ms);
+      if (bm_dfu_internal()) {
+        err = bm_dfu_host_get_chunk(flash_offset,
+                                    payload_header->chunk.payload_buf,
+                                    payload_len, flash_read_timeout_ms);
+      } else if (host_ctx.data_queue) {
+        //TODO: handle gracefully if cannot malloc
+        uint8_t *data_buf = (uint8_t *)bm_malloc(host_ctx.img_info.chunk_size);
+        err = bm_queue_receive(host_ctx.data_queue, data_buf,
+                               host_ctx.host_timeout_ms);
+        memcpy(payload_header->chunk.payload_buf, data_buf, payload_len);
+        bm_free(data_buf);
```

Comment:
I don't love this flow, but when adding data to a queue, you add the whole buffer allocated to that queue item, when receiving from a queue you get the whole buffer allocated to that queue item. If this were to apply to `payload_header->chunk.payload_buf`, we risk the chance of buffer overflow on the last chunk being queued up.

Reply:
This has been resolved by using stream buffers